File: testclient/z80.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;

namespace ProjectCambridge
{
    public partial class Z80
    {
        // The Z80 microprocessor user manual can be downloaded from Zilog directly, here:
        //    http://tinyurl.com/z80manual
        // 
        // Other useful details of the Z80 architecture can be found here:
        //    http://landley.net/history/mirror/cpm/z80.html
        // and here: 
        //    http://z80.info/z80code.htm

        // Memory and clock
        private Memory memory;

        public Z80(Memory memory)
        {
            this.memory = memory;
            this.Reset();
        }

        // operations
        private byte INC(byte reg)
        {
            fP = (reg == 0x7F);
            fH = (reg & 0xF) == 0xF;
            reg++;
            fZ = IsZero(reg);
            fS = IsSign(reg);
            fN = false;

            return reg;
        }

        private byte DEC(byte reg)
        {
            fP = (reg == 0x80);
            fH = false; // TODO: set if borrow from bit 4
            reg--;
            fZ = IsZero(reg);
            fS = IsSign(reg);
            fN = true;

            return reg;
        }

        private void SUB(byte reg)
        {
            a -= reg;
            fS = IsSign(reg);
            fZ = IsZero(reg);
        }

        private void CALL()
        {
            var callAddr = GetNextWord();

            pc += 2;
            PUSH(pc);

            pc = callAddr;
        }

        private void RST(byte addr)
        {
            PUSH(pc);
            pc = addr;
        }

        private void PUSH(ushort val)
        {
            memory.WriteByte(--sp, HighByte(val));
            memory.WriteByte(--sp, LowByte(val));
        }

        private ushort POP()
        {
            var lo = memory.ReadByte(sp++);
            var hi = memory.ReadByte(sp++);
            return (ushort)((hi << 8) + lo);
        }

        private byte RLC(byte reg)
        {
            // rotates register r to the left
            // bit 7 is copied to carry and to bit 0
            fC = IsSign(reg);
            reg <<= 1;
            if (fC) reg = (byte)SetBit(reg, 0);

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            return reg;
        }

        private byte RRC(byte reg)
        {
            fC = IsBitSet(reg, 0);
            reg >>= 1;
            if (fC) reg = (byte)SetBit(reg, 7);

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            return reg;

        }

        private byte RL(byte reg)
        {
            // rotates register r to the left, through carry. 
            // carry becomes the LSB of the new r
            bool bit0 = fC;

            fC = IsSign(reg);
            reg <<= 1;

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            if (bit0) reg = (byte)SetBit(reg, 0);

            return reg;
        }

        private byte RR(byte reg)
        {
            bool bit7 = fC;

            fC = IsBitSet(reg, 0);
            reg >>= 1;

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            if (bit7) reg = (byte)SetBit(reg, 7);

            return reg;
        }

        private byte SLA(byte reg)
        {
            fC = IsSign(reg);
            reg <<= 1;

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            return reg;
        }

        private byte SRA(byte reg)
        {
            bool bit7 = IsSign(reg);

            fC = IsBitSet(reg, 0);
            reg >>= 1;

            if (bit7) SetBit(reg, 7);

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            return reg;
        }

        private byte SLL(byte reg)
        {
            // technically, SLL is undocumented
            fC = IsBitSet(reg, 7);
            reg <<= 1;
            SetBit(reg, 1);

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            return reg;
        }

        private byte SRL(byte reg)
        {
            fC = IsBitSet(reg, 0);
            reg >>= 1;
            ResetBit(reg, 7);

            fS = IsSign(reg);
            fZ = IsZero(reg);
            fH = false;
            fP = IsParity(reg);
            fN = false;

            return reg;
        }

        // 
        private byte GetNextByte()
        {
            var byteRead = memory.ReadByte(pc);
            pc++;
            return byteRead;
        }

        private ushort GetNextWord()
        {
            var wordRead = memory.ReadWord(pc);
            pc += 2;
            return wordRead;
        }

        // interrupts

        public void Reset()
        {
            a = b = c = d = e = h = l = 0;
            f = 0;
            ix = iy = pc = sp = 0;
        }

        public string GetState()
        {
            // AF BC DE HL AF' BC' DE' HL' IX IY SP PC
            // I R IFF1 IFF2 IM < halted > < tstates >

            var state = $"A  {a:X2} | BC  {bc:X4} | DE  {de:X4} | HL  {hl:X4}\n";
            state += $"A' {a_:X2} | BC' {bc_:X4} | DE' {de_:X4} | HL' {hl_:X4}\n";
            state += $"IX {ix:X4} | IY {iy:X4} | SP {sp:X4} | PC {pc:X4}\n";

            state += "Flags: ";
            if (fC) state += "C";
            if (fN) state += "N";
            if (fP) state += "P";
            if (fH) state += "H";
            if (fZ) state += "Z";
            if (fS) state += "S";

            return state;
        }
    }
}

File: EmulatorCore/z80.dart
// z80.dart -- implements the Zilog Z80 processor core
//
// Reference notes:
// The Z80 microprocessor user manual can be downloaded from Zilog:
//    http://tinyurl.com/z80manual
//
// Other useful details of the Z80 architecture can be found here:
//    http://landley.net/history/mirror/cpm/z80.html
// and here:
//    http://z80.info/z80code.htm

class Z80 {
  // REGISTERS
  final flags = const {
    'C': 0x01, // carry flag (bit 0)
    'N': 0x02, // add/subtract flag (bit 1)
    'P': 0x04, // parity/overflow flag (bit 2)
    'F3': 0x08, // undocumented flag
    'H': 0x10, // half carry flag (bit 4)
    'F5': 0x20, // undocumented flag
    'Z': 0x40, // zero flag (bit 6)
    'S': 0x80 // sign flag (bit 7)
  };

  // Core registers
  int a, f, b, c, d, e, h, l;
  int ix, iy;

  // The alternate register set (A', F', B', C', D', E', H', L')
  int a_, f_, b_, c_, d_, e_, h_, l_;

  int i; // Interrupt Page Address register
  int r; // Memory Refresh register

  int pc; // Program Counter
  int sp; // Stack pointer

  bool iff1;
  bool iff2;

  int im; // Interrupt Mode

  int get af => a * 256 + f;
  set af(num value) {
    a = highByte(value);
    f = lowByte(value);
  }

  int get bc => b * 256 + c;
  set bc(num value) {
    b = highByte(value);
    c = lowByte(value);
  }

  int get de => d * 256 + e;
  set de(num value) {
    d = highByte(value);
    e = lowByte(value);
  }

  int get hl => h * 256 + l;
  set hl(num value) {
    h = highByte(value);
    l = lowByte(value);
  }

  int get ixh => ix & 0xFF00 >> 8;
  int get ixl => ix & 0x00FF;

  int get iyh => iy & 0xFF00 >> 8;
  int get iyl => iy & 0x00FF;

  bool get fC => f & flags['C'] == flags['C'];
  bool get fN => f & flags['N'] == flags['N'];
  bool get fP => f & flags['P'] == flags['P'];
  bool get fF3 => f & flags['F3'] == flags['F3'];
  bool get fH => f & flags['H'] == flags['H'];
  bool get fF5 => f & flags['F5'] == flags['F5'];
  bool get fZ => f & flags['Z'] == flags['Z'];
  bool get fS => f & flags['S'] == flags['S'];

  int highByte(int value) => (value & 0xFF00) >> 8;
  int lowByte(int value) => value & 0x00FF;
}
